app/src/main/java/org/schabi/newpipe/player/resolver/ExtractorDataSourceResolver.java

```diff
@@ -0,0 +1,321 @@
+        private ProgressiveVideoDataSpec() { }
+
+        public static Uri uriFrom(final int serviceId,
+                                  @NonNull final String streamUrl,
+                                  @NonNull final VideoStream serializable) {
+            return URI_RESOLVER.bundle(serviceId, streamUrl, serializable);
+        }
+
+        @NonNull
+        @Override
+        public String getSpecType() {
+            return "progressive-video";
+        }
+
+        @NonNull
+        @Override
+        public Class<VideoStream> getClassType() {
+            return VideoStream.class;
+        }
+
+        @NonNull
+        @Override
+        public Uri resourceUriFrom(@NonNull final StreamInfo info,
+                                   @NonNull final VideoStream serialized) {
+            final java.util.stream.Stream<List<VideoStream>> videos = java.util.stream.Stream.of(
+                    info.getVideoStreams(),
+                    info.getVideoOnlyStreams(),
+                    info.getSegmentedVideoStreams());
+
+            return videos
+                    .flatMap(List::stream)
+                    .filter(serialized::equalStats)
+                    .map(Stream::getUrl)
+                    .map(Uri::parse)
+                    .findFirst()
+                    .orElseThrow(NoSuchElementException::new);
+        }
+    }
+
+    public static final class LiveStreamDataSpec implements ExtractorDataSpec<String> {
+        private static final LiveStreamDataSpec URI_RESOLVER = new LiveStreamDataSpec();
+        private static final String DASH_PROTOCOL = "dash";
+        private static final String HLS_PROTOCOL = "hls";
+        private LiveStreamDataSpec() { }
+
+        public static Uri dashUriFrom(final int serviceId, @NonNull final String streamUrl) {
+            return URI_RESOLVER.bundle(serviceId, streamUrl, DASH_PROTOCOL);
+        }
+
+        public static Uri hlsUriFrom(final int serviceId, @NonNull final String streamUrl) {
+            return URI_RESOLVER.bundle(serviceId, streamUrl, HLS_PROTOCOL);
+        }
+
+        @NonNull
+        @Override
+        public String getSpecType() {
+            return "live";
+        }
+
+        @NonNull
+        @Override
+        public Class<String> getClassType() {
+            return String.class;
+        }
+
+        @NonNull
+        @Override
+        public Uri resourceUriFrom(@NonNull final StreamInfo info, @NonNull final String protocol) {
+            final Uri manifestUri;
+            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
+                manifestUri = Uri.parse(info.getDashMpdUrl());
+            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
+                manifestUri = Uri.parse(info.getHlsUrl());
+            } else {
+                throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
+            }
+            return manifestUri;
```

Comment:
subjective. I actually prefer this coding style with preferably one return statement or an Exception (which is done here) or at least as few as possible return statements if the code would otherwise get ugly.

Reply:
Sorry but I think that 
```java
            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
                return Uri.parse(info.getDashMpdUrl());
            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
                return Uri.parse(info.getHlsUrl());
            }
            throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
```
is a lot better readable then:
```java
            final Uri manifestUri;
            if (DASH_PROTOCOL.equals(protocol) && info.getDashMpdUrl() != null) {
                manifestUri = Uri.parse(info.getDashMpdUrl());
            } else if (HLS_PROTOCOL.equals(protocol) && info.getHlsUrl() != null) {
                manifestUri = Uri.parse(info.getHlsUrl());
            } else {
                throw new NoSuchElementException("Protocol [" + protocol + "] not found.");
            }
            return manifestUri;
```

Shorter, less expressions, less ``{ }``, same functionality.